pkg/multicast/multicast.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicast
+
+import (
+	"net"
+
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"golang.org/x/net/ipv6"
+)
+
+var (
+	// v6Socket is the udp socket used to join/leave a multicast group
+	v6Socket net.PacketConn
+
+	mutex lock.Mutex
```

Comment:
The way that `mutex` is used here is more appropriate to use `sync.Once()` instead.

Reply:
If the init fails, it has another opportunity to init again.

Reply:
Good point. What kind of failures do you anticipate that will change over time?

If it's always EPERM or EADDRINUSE then it probably won't change from one attempt to the next. But if there's another common error case you're thinking of then leaving as-is could make sense.

Reply:
btw, I changed it to `net.ListenPacket("udp6", "[::]:0")`. I don't have any specific possible failure in mind. I agree this is unlikely to fail. Probably no ephemeral ports available? but this is ***extremely*** unlikely. I am okay either way, actually.

Reply:
If it's super unlikely, I think the best we can do is put a big fat warning message and just let the user know.

Reply:
it is logged as warning now, but not sure it is big or fat.  `"Failed to listen on socket for multicast: %v"` make it uppercase?